crates/accelerate/src/target_transpiler/mod.rs

```diff
@@ -0,0 +1,1285 @@
+    ///         documentation for the :class:`~qiskit.transpiler.Target` class).
+    /// Raises:
+    ///     AttributeError: If gate is already in map
+    ///     TranspilerError: If an operation class is passed in for ``instruction`` and no name
+    ///         is specified or ``properties`` is set.
+    #[pyo3(signature = (instruction, name, properties=None))]
+    fn add_instruction(
+        &mut self,
+        instruction: TargetOperation,
+        name: String,
+        properties: Option<PropsMap>,
+    ) -> PyResult<()> {
+        if self.gate_map.contains_key(&name) {
+            return Err(PyAttributeError::new_err(format!(
+                "Instruction {:?} is already in the target",
+                name
+            )));
+        }
+        let mut qargs_val: PropsMap = PropsMap::default();
+        match instruction {
+            TargetOperation::Variable(_) => {
+                qargs_val = IndexMap::from_iter([(None, None)].into_iter());
+                self.variable_class_operations.insert(name.clone());
+            }
+            TargetOperation::Normal(_) => {
+                if let Some(properties) = properties {
+                    let inst_num_qubits = instruction.num_qubits();
+                    if properties.contains_key(&None) {
+                        self.global_operations
+                            .entry(inst_num_qubits)
+                            .and_modify(|e| {
+                                e.insert(name.clone());
+                            })
+                            .or_insert(HashSet::from_iter([name.clone()]));
+                    }
+                    for qarg in properties.keys() {
+                        let mut qarg_obj = None;
+                        if let Some(qarg) = qarg {
+                            if qarg.len() != inst_num_qubits as usize {
+                                return Err(TranspilerError::new_err(format!(
+                                    "The number of qubits for {name} does not match\
+                                    the number of qubits in the properties dictionary: {:?}",
+                                    qarg
+                                )));
+                            }
+                            self.num_qubits =
+                                Some(self.num_qubits.unwrap_or_default().max(
+                                    qarg.iter().fold(0, |acc, x| {
+                                        if acc > x.index() {
+                                            acc
+                                        } else {
+                                            x.index()
+                                        }
+                                    }) + 1,
+                                ));
+                            qarg_obj = Some(qarg.clone())
+                        }
+                        qargs_val.insert(qarg_obj.to_owned(), properties[qarg].clone());
```

Comment:
This clone seems a bit weird, since we own `properties` I would expect us to be able to just move the data into `gate_map` directly.

Reply:
I will use `IndexMap::swap_remove()` to avoid cloning.